pkg/schema/definitions/handler.go

```diff
@@ -0,0 +1,197 @@
+package definitions
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/rancher/apiserver/pkg/apierror"
+	"github.com/rancher/apiserver/pkg/types"
+	"github.com/rancher/steve/pkg/schema/converter"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+var (
+	internalServerErrorCode = validation.ErrorCode{
+		Status: http.StatusInternalServerError,
+		Code:   "InternalServerError",
+	}
+	notRefreshedErrorCode = validation.ErrorCode{
+		Status: http.StatusServiceUnavailable,
+		Code:   "SchemasNotRefreshed",
+	}
+)
+
+// schemaDefinitionHandler is a byID handler for a specific schema, which provides field definitions for all schemas.
+// Does not implement any method allowing a caller to list definitions for all schemas.
+type schemaDefinitionHandler struct {
+	sync.RWMutex
+
+	// lastRefresh is the last time that the handler retrieved models from kubernetes.
+	lastRefresh time.Time
+	// refreshStale is the duration between lastRefresh and the next refresh of models.
+	refreshStale time.Duration
+	// client is the discovery client used to get the groups/resources/fields from kubernetes.
+	client discovery.DiscoveryInterface
+	// models are the cached models from the last response from kubernetes.
+	models *proto.Models
+	// schemaToModel is a map of the schema name to the model for that schema. Can be used to load the
+	// top-level definition for a schema, which can then be processed by the schemaFieldVisitor.
+	schemaToModel map[string]string
+}
+
+// byIDHandler is the Handler method for a request to get the schema definition for a specifc schema. Will use the
+// cached models found during the last refresh as part of this process.
+func (s *schemaDefinitionHandler) byIDHandler(request *types.APIRequest) (types.APIObject, error) {
+	// pseudo-access check, designed to make sure that users have access to the schema for the definition that they
+	// are accessing.
+	requestSchema := request.Schemas.LookupSchema(request.Name)
+	if requestSchema == nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.NotFound, "no such schema")
+	}
+
+	if s.needsRefresh() {
+		err := s.refresh()
+		if err != nil {
+			logrus.Errorf("error refreshing schemas %s", err.Error())
+			return types.APIObject{}, apierror.NewAPIError(internalServerErrorCode, "error refreshing schemas")
+		}
+	}
+
+	// lock only in read-mode so that we don't read while refresh writes. Only use a read-lock - using a write lock
+	// would make this endpoint only usable by one caller at a time
+	s.RLock()
+	defer s.RUnlock()
+
+	if s.models == nil {
```

Comment:
Strictly speaking the case isn't possible. However, I don't think it's a good practice to have a function rely on internal behavior of another function. In this case, since refresh doesn't return a value I don't think `byIDHandler` can assume that this value is filled in. Because of that, I would prefer to keep this as-is.

Reply:
IIUC then this is a case of "defensive programming" and I dislike it in general, as it adds unnecessary complexity/code/checks (here, in callers that will end up expecting one more error, when code will inevitably be used by other programmers who do not necessarily know all the details...).

I will not insist on this being changed for the purposes of this PR.

Reply:
There's a few ways that this could be characterized. Personally, I'd lean toward not assuming anything that isn't part of the functional contract.

Since we intend to make refresh no longer a part of the request handling logic (i.e. the Handler will need to have a refresh initiated by another source), I think that this will have a more direct value in the future. For that reason. I'm going to leave the code as-is and resolve the conversation. If you disagree, please un-resolve with further feedback.